### ISSUE TITLE
Authorize Settings In swagger-ui Not Working When Using Oauth2 Security Definitions

### ISSUE THREAD

**Issue by NColey** [rswag/rswag#144]
I have an api only rails 5 app, that's using doorkeeper + devise with the Oauth2 resource owner password grant flow. I'm also using swagger for my tests. When I went to generate the ui I noticed something strange - the Authorize button didn't have any available authorizations (the Authorize button within the popup also didn't work).

FWIW the tests pass (I will include them below). I've been googling for a few hours trying to figure out where exactly I went wrong with setting things up (and editing this issue as I get better info and update my tests) but it's starting to look like it may be an issue with how swagger-ui generates docs for people using the oauth password grant flow. It's possible this exists for other flows as well - I haven't tried that as those aren't what my api is using.

It's more than possible that the issue may be with my implementation of the security definitions for oauth, or how I'm structuring things in the swagger helper, or even my expectation of what I should see in that authorize popup based on examples I've seen online -  but I think its likely enough that it's a swagger-ui bug that I wanted to flag it just in case.

For reference below is my spec, the swagger helper, and a screenshot of the issue in swagger-ui.

user_controller_spec.rb

```
require "swagger_helper"


RSpec.describe "Api::V1::Users", type: :request do
  path "/v1/users/{id}" do
    get "Retrieves a User" do
      tags "Users"
      produces "application/vnd.api+json"
      parameter name: :id, in: :path, type: :integer
      parameter name: :Authorization, in: :header, type: :string, required: :true
      security [ oauth: [] ]

      response "200", "user found" do
        schema RswagHelper.json_api_schema(
          first_name: { type: :string },
          last_name: { type: :string },
          email: { type: :string },
          phone: { type: :string, 'x-nullable': true },
          country_code: { type: :string },
          date_of_birth: { type: :string, 'x-nullable': true },
          is_member: { type: :boolean },
          is_facebook_connected: { type: :boolean },
          is_amazon_connected: { type: :boolean }
        )

        let(:user) { FactoryBot.create(:user) }
        let(:token) { FactoryBot.create(:access_token, resource_owner_id: user.id) }
        let(:id) { user.id }
        let(:Authorization) { "Bearer #{token.token}" }

        run_test!
      end
    end
  end
end

```

For reference RswagHelper.json_api_schema is just: 
we have some helpers for json blobs that are repeated so they were consolidated into helpers. Removing that didn't affect this issue.
```
  def json_api_schema(attributes)
    {
      type: :object,
      properties: {
        data: {
          type: :object,
          properties: {
            id: { type: :string },
            type: { type: :string },
            attributes: {
              type: :object,
              properties: attributes,
              required: attributes.keys
            }
          },
          required: [ "id", "type", "attributes" ]
        }
      },
      required: [ "data" ]
    }
  end
```

swagger_helper.rb (just the relevant config)

```
    config.swagger_docs = {
    "v1/swagger.json" => {
      swagger: "2.0",
      info: {
        title: "User API V1",
        version: "v1"
      },
      paths: {},
      securityDefinitions: {
        oauth: {
          type: :oauth2,
          flows: {
            password: {
              tokenUrl: "https://localhost:3000/oauth/token"
            }
          }
        }
      }
    }
  }
```

Image of the authorize popup in swagger-ui.
<img width="1434" alt="swagger-ui" src="https://user-images.githubusercontent.com/8762012/44604787-0e8f6d80-a7b5-11e8-874e-b0e990dd60ba.png">

Please let me know if any of this was unclear or if there is additional information I can find. Thanks!

**Comment by domaindrivendev** [comment]
The structure of your `securityDefinitions` looks incorrect to me. It's worth noting that rswag currently uses Swagger 2.0. Here's the relevant section of the spec for reference:

https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#securityDefinitionsObject

In addition to the structural changes, you'll need to define one or more scopes and then for each operation, indicate which of those scopes are required. For example:

```ruby
# swagger_helper.rb
securityDefinitions: {
  oauth: {
    type: :oauth2,
    flows: "password",
    tokenUrl: "https://localhost:3000/oauth/token",
    scopes: [ "read", "write" ]
  }
}

# user_controller_spec.rb
get "Retrieves a User" do
  ...
  security [ oauth: [ "read" ] ]
```

**Comment by NColey** [comment]
@domaindrivendev you're 1000% right about the structure of securityDefinitions, I must have been misunderstanding what I was seeing in the spec (I was referencing [this](https://swagger.io/specification/#security-definitions-object-109)). I fixed that and it's now giving me authorization options. Thanks so much!

If it's not too much I have one remaining question - I'm encountering a new error now, in that when I input my email and password (our auth setup expects email and password vs username, but not sure that's the issue) I get the following error `Auth ErrorTypeError: Failed to fetch` - is there another piece of the puzzle I'm missing to get this working in the UI? For reference my updated `securityDefinitions` are:

```
config.swagger_docs = {
    "v1/swagger.json" => {
      swagger: "2.0",
      info: {
        title: "User API V1",
        version: "v1"
      },
      paths: {},
      securityDefinitions: {
        oauth: {
          type: :oauth2,
          flow: "password",
          tokenUrl: "https://localhost:3000/oauth/token",
          scopes: ["public"]
        }
      }
    }
  }
```

and this is a screenshot of the error, email address edited (for context I selected the basic auth option as it was the only one to not require a client_id and secret, we use the password grant flow as sending over a client_secret isn't appropriate for the use case of this api):

<img width="1435" alt="typeerror" src="https://user-images.githubusercontent.com/8762012/44612786-4b6d5b80-a7d9-11e8-810d-03a00cb9e32b.png">

Thanks again!
